### PR TITLE
MAGECLOUD-4168: Provide an Option to Specify Custom URL for Magento

### PR DESCRIPTION
**Pull request by oshmyheliuk**
### Description
Added possibility to configure hostname and port.

### Fixed Issues (if relevant)
https://magento2.atlassian.net/browse/MAGECLOUD-4168

### Manual testing scenarios
1. Launch Docker environment https://devdocs.magento.com/cloud/docker/docker-mode-production.html
2. Run build compose command with configured host `./vendor/bin/ece-docker build:compose --host=magento2.test`
3. After installation of Magento, check that site is accessible by `magento2.test` url not `magento2.docker`
4. Run `docker-compose down -v`
5. Launch Docker environment with configured port `./vendor/bin/ece-docker build:compose --host=magento2.test --port=8080`
6. Check that site is accessible by `magento2.test:8080` url not `magento2.docker`

releasenotes
Added the capability to configure the hostname and port in a local Docker environment using the `--host` and `--port` options in the `bin/magento2 build:compose` command. For example: `build:compose --host=magento2.local --port=8080`

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages

### CODE REVIEW

File: src/Config/Dist/Generator.php [hunk 1]
```diff
@@ -172,4 +153,39 @@ private function saveConfigEnv(string $filePath, array $config): void
...
+    /**
+     * Returns base configuration
+     *
+     * @param Config $config
+     * @return array
+     * @throws ConfigurationMismatchException
+     */
+    private function getBaseConfig(Config $config): array
+    {
+        $host = $config->get(SourceInterface::CONFIG_HOST) ?? Manager::DEFAULT_HOST;
```
Comment (shiftedreality) [inline src/Config/Dist/Generator.php]: `SourceInterface` should not be used outside of `Config`, please add public methods to `Config`

File: src/Compose/Manager.php [hunk 2]
```diff
@@ -12,7 +12,8 @@
  */
 class Manager
 {
-    public const DOMAIN = 'magento2.docker';
+    public const DEFAULT_HOST = 'magento2.docker';
```
Comment (shiftedreality) [inline src/Compose/Manager.php]: Let's move this to `BaseSource` to be the default value

**Review by shiftedreality — CHANGES_REQUESTED** [review]

File: src/Compose/Manager.php (hunk 2 above)
Comment (oshmyheliuk) [reply src/Compose/Manager.php]: This constant was here, I just renamed it.
I don't think that `BaseSource` is a good place for them. I'll move contstants and related methods to the `Config` class.

Comment (shiftedreality) [reply src/Compose/Manager.php]: These are default values, which can be changed by any source, `BaseSource` should be correct

File: src/Config/Dist/Generator.php (hunk 1 above)
Comment (oshmyheliuk) [reply src/Config/Dist/Generator.php]: done

File: src/Compose/Manager.php (hunk 2 above)
Comment (oshmyheliuk) [reply src/Compose/Manager.php]: done

File: src/Command/BuildCompose.php [hunk 3]
```diff
@@ -204,6 +204,18 @@ protected function configure(): void
                 null,
                 InputOption::VALUE_OPTIONAL,
                 'Cloud environment variables'
+            )
+            ->addOption(
+                Source\CliSource::OPTION_HOST,
+                null,
+                InputOption::VALUE_OPTIONAL,
+                'Host name'
```
Comment (shiftedreality) [inline src/Command/BuildCompose.php:212]: @meker12 we probably need some task to review all customer-facing texts in Cloud Docker

File: src/Config/Config.php [hunk 4]
```diff
@@ -292,4 +292,24 @@ public function getVariables(): array
...
     }
+
+    /**
+     * Returns host value or default if host not set
+     *
+     * @return string
+     */
+    public function getHost(): string
+    {
+        return $this->get(SourceInterface::CONFIG_HOST) ?? BaseSource::DEFAULT_HOST;
```
Comment (shiftedreality) [inline src/Config/Config.php]: You don't need ternary here, it must be always by default what you set in `BaseSource`.

Config must not know about sources

**Review by shiftedreality — CHANGES_REQUESTED** [review]

File: src/Config/Config.php (hunk 4 above)
Comment (oshmyheliuk) [reply src/Config/Config.php]: I didn't set anything in BaseSource, just added constants

Comment (oshmyheliuk) [reply src/Config/Config.php]: Updated

**Review by shiftedreality — APPROVED** [review]

**Review by arhiopterecs — APPROVED** [review]

**Comment by andriyShevtsov** [comment]
QA approved

**Pull request merged**